### PR TITLE
attempt port of goon process scheduler.

### PR DESCRIPTION
**Pull request by eswordthecat**
- [ ] it's 2016 now?
- [x] add master controller
  - [x] mobs
  - [x] diseases
  - [x] nanouis
  - [x] machines
  - [x] powernet
  - [x] air
  - [x] pipenet
  - [x] events
  - [x] vote
  - [x] sun
  - [x] ticker
  - [x] garbage
  - [x] remove self to running controllers
- [x] add shuttle controllers
  - [x] supply
  - [x] emergency
  - [x] remove self to running controllers
- [x] add ligthing controller
  - [x] remove self to running controllers
- [x] failsafe controller is kill
  - [x] remove self to running controllers
- [ ] debugging is yay
  - [ ] removed old debuggers for old controllers
  - [ ] add each to process
  - [ ] add it to the main scheduler
- [ ] tested
  - [ ] nothing breaks ('no')
- [ ] mergeable now?
  - [ ] updated with the current code at Bleeding-Edge branch
  - [ ] introduced to player community
  - [ ] address everyone concerns
  - [ ] squashed commits

### CODE REVIEW

File: code/controllers/_stubs.dm [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	world << msg
+*/
+/**
+ * logTheThing
+ *
+ * In goonstation, this proc writes a message to either the world log or diary.
+ *
+ * Blame Keelin.
+ */
+/proc/logTheThing(type, source, target, text, diaryType)
```
Comment (Aurx) [inline code/controllers/_stubs.dm]: What's wrong with the existing logging functions?

Comment (eswordthecat) [inline code/controllers/_stubs.dm]: Nothing, that's a part of the released code so maybe i'll remove it later.

**Comment by IratePirate** [comment]
What benefits will this bring out of curiosity?

**Comment by volundr-** [comment]
The overall effect is to allow the master controller processes to run concurrently, and allow other stuff to occur while the master controller is running. The old master controller blocks execution of other procs in a lot of cases, causing rather punishing lag.

**Comment by volundr-** [comment]
FYI the scheduler has been relicensed under CC-BY-NC

**Comment by volundr-** [comment]
Also I fixed a bug with timing in it just now so you'll want to pull the changes.

**Comment by eswordthecat** [comment]
updooted!

**Comment by volundr-** [comment]
I removed the stuff referencing goonstation's ticker process -- completely decoupled now.

**Comment by eswordthecat** [comment]
thank you

**Comment by volundr-** [comment]
FYI I fixed another issue - merge changes from upstream

**Comment by eswordthecat** [comment]
yep, merging it

**Comment by wild-billy** [comment]
Not bad. :+1:x:100: for the library structure, AND from Goon to boot. Cooperation is the way to go. We'll be doing quite a bit more of that soon (tm).

Also, may want to note that @Carn is also working on this (see https://github.com/tgstation/-tg-station/pull/6706).

And I'd say something about muh OOP or muh concurrency, but the teaching/preaching just isn't getting through on this topic. Looks like this is one of those things that just gotta be learned the hard way.

**Comment by optimumtact** [comment]
Carn's rewrite is a bit different, since it just moduralises the old MC loop we were using

**Comment by optimumtact** [comment]
Also I'm pretty sure CC BY NC is incompatible with GPL code

**Comment by d3athrow** [comment]
@optimumtact 
I believe he changed the license after consulting his goon lawyer guy, atleast thats the gist i got after skimming IRC when he made the change.

**Comment by ririchiyo** [comment]
Who actually owns the rights to the original license?

**Comment by optimumtact** [comment]
Volundr does

**Comment by ririchiyo** [comment]
I meant specifically for baystation/tgstation.

**Comment by Jalleo** [comment]
Whoever runs the baystation or tg station at that point in time.

**Comment by tkdrg** [comment]
Not really. The license to each piece of code belongs to each person who contributed it. That is why we can't change the license of old code, but we can require newly contributed code to have the new license.

**Comment by JamieH** [comment]
out of curiosity, what kinda performance boost did you get out of implementing goons MC over the old MC?

**Comment by ririchiyo** [comment]
When we ran out test from it, there was little lag when powernets were being destroyed and then recreated. However, when meteors were hitting the station and dealing with the environment there was some noticeable lag that quickly passed but this likely wasn't because of the MC.

It was a huge improvement overall, though.

**Comment by Iamgoofball** [comment]
what

**Pull request closed without merging**